### PR TITLE
fix begin trasaction after reconnect

### PR DESCRIPTION
**Pull request by kalinin-k-a**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | 

#### Summary

This is a small fix if we lose the connection and work with transactions after reconnect.
The code examples below demonstrate a problem:

```php
/** Doctrine\DBAL\Connection $co */
$co->executeQuery('CREATE TABLE if not exists test11(test int not null)');
$co->executeQuery('truncate table test11');
$co->executeQuery('insert into test11 values (1)');
$co->beginTransaction();
$co->executeQuery('insert into test11 values (2)');
$v = $co->fetchAll('select * from test11');
$co->close(); // here we lost a connection or closed it for some reason
$co->beginTransaction(); /* p1 */
$co->executeQuery('insert into test11 values (33)');
$co->rollback(); /* p2 */
$v = $co->fetchAll('select * from test11'); // I expect the value 33 will not be selected from the table because it was rollbacked. But the transaction was not started on the line "p1" because of nestingLevel had not beed reset after open new connection.
```

### CODE REVIEW

**Comment by kalinin-k-a** [comment]
This is a small fix if we lose the connection and work with transactions after reconnect.
The code examples below demonstrate a problem:
/** Doctrine\DBAL\Connection $co */
$co->executeQuery('CREATE TABLE if not exists test11(test int not null)');
         $co->executeQuery('truncate table test11');
         $co->executeQuery('insert into test11 values (1)');
         $co->beginTransaction();
         $co->executeQuery('insert into test11 values (2)');
         $v = $co->fetchAll('select * from test11');
         $co->close(); // here we lost a connection or closed it for some reason
         $co->beginTransaction(); /* p1 */
         $co->executeQuery('insert into test11 values (33)');
         $co->commit(); /* p2 */
        $v = $co->fetchAll('select * from test11'); // I expect to see here value of 33 in the table, but it was not commited because transaction nesting level had not been reset after reconnect

**Comment by morozov** [comment]
@kalinin-k-a please turn the scenario on the description into a test.

**Comment by kalinin-k-a** [comment]
> 
> 
> @kalinin-k-a please turn the scenario in the description into a test.

done

**Comment by greg0ire** [comment]
Why are there 2 PRs for this? #3679 looks the same, but on another branch and without tests, which one is the right one?

**Comment by greg0ire** [comment]
Please squash your commits

> fix begin trasaction ofter reconnect

Please explain in your commit message what was right, and why your changes fix that situation. Maybe

```
Reset transaction nesting level on connection loss

When the connection is lost, subsequent transactions will no longer be nested because they are executed in a brand new session. Our internal representation of the nesting should take this into account.
```

**Comment by kalinin-k-a** [comment]
> 
> 
> Why are there 2 PRs for this? #3679 looks the same, but on another branch and without tests, which one is the right one?

Another is into v2.5. I'll also write tests

**Review by greg0ire — DISMISSED** [review]

**Comment by greg0ire** [comment]
Please associate your email address with your Github account, or change the email in your commits to an address already associated with it.

**Comment by kalinin-k-a** [comment]
commits are squashed, email is changed. Thank you for reminding !

**Review by greg0ire — DISMISSED** [review]

**Comment by kalinin-k-a** [comment]
pushed again because mysql server had gone away during functional tests previously

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 1]
```diff
@@ -72,6 +72,18 @@ public function testTransactionNestingBehavior() : void
...
+    public function testTransactionNestingLevelIsResetOnReconnect() : void
+    {
+        $this->connection->beginTransaction();
+        $this->connection->beginTransaction();
+        self::assertEquals(2, $this->connection->getTransactionNestingLevel());
+        $this->connection->close(); // connection is lost
+        $this->connection->beginTransaction(); // should connect, reset nesting level and increase it once
+        self::assertEquals(1, $this->connection->getTransactionNestingLevel());
+        $this->connection->commit();
+        self::assertEquals(0, $this->connection->getTransactionNestingLevel());
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Please use the scenario from the description. The test as it's written now does not reflect the actual issue.

Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: I’m not sure that scenario from description is good for test. It just shows one case that may be affected by nesting level issue. Also the result from the description may happen if for example commit works wrong.

I fixed nesting level reset, so i tested the same.

Do you insist on change test?

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Without that case, it's not clear why we want to reset the transaction nesting level after establishing a connection. The level is an internal implementation detail of the DBAL, it doesn't have to be tested in a functional way but I'd rather have a functional case which relies on it. Even if it's just one case.

Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Ok.
Strange, but Insert statement fails the test now with "Table not found". Seems like table exist only before closing connection in testing environment. I'll look again a little later.

Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Fixed. It was sqlite connection with memory=true. Please review once more

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 2]
```diff
@@ -70,6 +70,40 @@ public function testTransactionNestingBehavior() : void
             $this->connection->rollBack();
             self::assertEquals(0, $this->connection->getTransactionNestingLevel());
         }
+
+        $this->connection->beginTransaction();
+        $this->connection->close();
+        $this->connection->beginTransaction();
+        self::assertEquals(1, $this->connection->getTransactionNestingLevel());
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Does this test need to be updated or it's better to put this in a separate test?

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 3]
```diff
@@ -70,6 +70,40 @@ public function testTransactionNestingBehavior() : void
...
+            $connection = $this->connection;
+        }
+
+        $connection->executeQuery('CREATE TABLE test_nesting(test int not null)');
+
+        $this->connection->beginTransaction();
+        $this->connection->beginTransaction();
+        $connection->close(); // connection closed in runtime (for example if lost or another application logic)
+
+        $connection->beginTransaction(); // should connect, reset nesting level and increase it once
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: The comment doesn't seem relevant to what the test is about.

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 2 above)
Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: The name of this test is very common  - testTransactionNestingBehavior(), and my fix affects transaction nesting behavior, so I uppdate this)

I can move this lines into a separate test and rename this to something

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Please do.

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: "Should realy start the transaction, and the rollback() call below realy leads to rollback it "

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 4]
```diff
@@ -70,6 +70,40 @@ public function testTransactionNestingBehavior() : void
...
+        $connection->executeQuery('insert into test_nesting values (33)');
+        $connection->rollback();
+
+        self::assertEquals(0, $connection->fetchColumn('select count(*) from test_nesting'));
```
Comment (SenseException) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Should the transaction nesting level be checked here too?

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (SenseException) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: The transaction nesting level could be checked here too.

**Review by SenseException — DISMISSED** [review]

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: > "Should realy start the transaction, and the rollback() call below realy leads to rollback it " - better?

Just remove the comment. The previous one says enough.

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 4 above)
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Not really. The fact that the transaction was successfully rolled back is enough. The nesting level is an implementation detail in this case.

Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: > 
> 
> Should the transaction nesting level be checked here too?

The exact values of nestingLevel is completely tested in testTransactionNestingBehavior()

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Ok,. done

**Review by morozov — DISMISSED** [review]

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Hm… it's still there.

Comment (kalinin-k-a) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Hm, it really was. Forgot to commit, sorry. Now it's realy done)

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thank you @kalinin-k-a.
